File: messagematch-typescript/src/match/pathmatcher.test.ts
import {expect, describe, test} from "vitest";
import {PathMatcher} from "./PathMatcher";

describe('pathmatcher (TS)', () => {
    test('should match a simple path', () => {
        const matcherPath:string = "/storedNumbers";
        const concretePath:string = "/storedNumbers";
        const pathMatcher = new PathMatcher(matcherPath, concretePath, {});
        expect(pathMatcher.matches()).toBeTruthy();
    });

    test('should match Int Path Param', () => {
        const matcherPath:string = "/{$Int=accountId,5}/storedNumbers";
        const concretePath:string = "/7/storedNumbers";
        const pathMatcher = new PathMatcher(matcherPath, concretePath, {});
        expect(pathMatcher.matches()).toBeTruthy();
        expect(pathMatcher.binding["accountId"]).toBe("7");
    });

    test('should match complex Matcher and bind', () => {
        const matcherPath:string = "/travel/{$String=from}-{$String=to}";
        const concretePath:string = "/travel/London-Edinburgh";
        const pathMatcher = new PathMatcher(matcherPath, concretePath, {});
        expect(pathMatcher.matches()).toBeTruthy();
        expect(pathMatcher.binding["from"]).toBe("London");
        expect(pathMatcher.binding["to"]).toBe("Edinburgh");
    });

    test('should not match', ()=>{
        const matcherPath:string = "/travel/{$String=from}-{$String=to}";
        const concretePath:string = "/traveling/London-Edinburgh";
        const pathMatcher = new PathMatcher(matcherPath, concretePath, {});
        expect(pathMatcher.matches()).toBeFalsy();
    })

    test('should respect existing bindings', ()=>{
        const matcherPath:string = "/travel/{$String=from}-{$String=to}";
        const concretePath:string = "/travel/London-Edinburgh";

        const pathMatcher = new PathMatcher(matcherPath, concretePath, {"from": "Bristol"});
        expect(pathMatcher.matches()).toBeFalsy();
    })

    test('should match whole path', ()=>{
        const matcherPath:string = "/storedNumbers";
        const concretePath:string = "/storedNumbersBroken";
        const pathMatcher = new PathMatcher(matcherPath, concretePath, {});
        expect(pathMatcher.matches()).toBeFalsy();
    })
});

File: messagematch/src/test/java/org/forwoods/messagematch/match/PathMatcherTest.java
package org.forwoods.messagematch.match;

import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class PathMatcherTest
{
    @Test
    public void shouldMatchSimple()
    {
        String matcherPath = "/storedNumbers";
        String concrete = "/storedNumbers";
        final PathMatcher pathMatcher = new PathMatcher(matcherPath, concrete, Map.of("from", "Bristol"));
        assertTrue(pathMatcher.matches());
    }

    @Test
    public void shouldMatchIntPathParam()
    {
        String matcherPath = "/{$Int=accountId,5}/storedNumbers";
        String concrete = "/7/storedNumbers";
        final PathMatcher pathMatcher = new PathMatcher(matcherPath, concrete, Map.of("from", "Bristol"));
        assertTrue(pathMatcher.matches());
        assertEquals("7", pathMatcher.getBindings().get("accountId"));
    }
    @Test
    public void shouldMatchComplexMatcher()
    {
        String matcherPath = "/travel/{$String=from}-{$String=to}";
        String concrete = "/travel/London-Edinburgh";
        final PathMatcher pathMatcher = new PathMatcher(matcherPath, concrete, Map.of());
        assertTrue(pathMatcher.matches());
        assertEquals("London", pathMatcher.getBindings().get("from"));
        assertEquals("Edinburgh", pathMatcher.getBindings().get("to"));
    }

    @Test
    public void shouldNotMatch()
    {
        String matcherPath = "/travel/{$String=from}-{$String=to}";
        String concrete = "/traveling/London-Edinburgh";
        final PathMatcher pathMatcher = new PathMatcher(matcherPath, concrete, Map.of());
        assertFalse(pathMatcher.matches());
    }

    @Test
    public void shouldRespectExistingBindings()
    {
        String matcherPath = "/travel/{$String=from}-{$String=to}";
        String concrete = "/travel/London-Edinburgh";
        final PathMatcher pathMatcher = new PathMatcher(matcherPath, concrete, Map.of("from", "Bristol"));
        assertFalse(pathMatcher.matches());
    }

    @Test
    public void shouldMatchWholePath()
    {
        String matcherPath = "/storedNumbers";
        String concrete = "/storedNumbersBroken";
        final PathMatcher pathMatcher = new PathMatcher(matcherPath, concrete, Map.of());
        assertFalse(pathMatcher.matches());
    }
}
